### PR TITLE
Add Family entity with additional configuration

### PR DESCRIPTION
**Pull request by ieatkillerbees**
Closes [ch18]

- Move bootstrap out of html/index.php to reuse it for CLI
- Set up doctrine CLI
- Create Family and Schedule (stub) entities with Test setup
- Configure Doctrine Migrations

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -1,14 +1,14 @@
...
-DB_CONNECTION=mysql
-DB_HOST=127.0.0.1
-DB_PORT=3306
-DB_DATABASE=choredo
-DB_USERNAME=choredo
-DB_PASSWORD=secret
-
-CACHE_DRIVER=file
-QUEUE_DRIVER=sync
+-APP_ENV=local
```
Comment (greydnls) [inline .env.example]: don't c/p off of git diffs... it adds `-` when they shouldn't be there.

File: cli-config.php [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    new \Doctrine\DBAL\Migrations\Tools\Console\Command\ExecuteCommand(),
+    new \Doctrine\DBAL\Migrations\Tools\Console\Command\GenerateCommand(),
+    new \Doctrine\DBAL\Migrations\Tools\Console\Command\LatestCommand(),
+    new \Doctrine\DBAL\Migrations\Tools\Console\Command\MigrateCommand(),
+    new \Doctrine\DBAL\Migrations\Tools\Console\Command\StatusCommand(),
+    new \Doctrine\DBAL\Migrations\Tools\Console\Command\VersionCommand(),
+    new \Doctrine\DBAL\Migrations\Tools\Console\Command\DiffCommand(),
+];
+
+$application->addCommands($commands);
```
Comment (greydnls) [inline cli-config.php:60]: I think you need to `array_merge($migrationCommands, $commands)` otherwise that `$migrationCommands` var is unused

File: composer.json [hunk 3]
```diff
@@ -8,9 +8,10 @@
         "php": ">=7.0",
         "laravel/lumen-framework": "5.5.*",
         "vlucas/phpdotenv": "~2.2",
-        "laravel-doctrine/orm": "1.3.*",
         "ramsey/uuid": "^3.7",
-        "ramsey/uuid-doctrine": "^1.4"
+        "ramsey/uuid-doctrine": "^1.4",
+        "doctrine/orm": "^2.5",
```
Comment (greydnls) [inline composer.json]: `--sort-packages`, pls.

File: docker-compose.yml [hunk 4]
```diff
@@ -16,7 +16,7 @@ services:
   db:
     image: postgres:10.0
     ports:
-      - "3306:3306"
+      - "5432:5432"
```
Comment (greydnls) [inline docker-compose.yml:19]: yer ops dude sucks

File: migrations/Version20171023043854.php [hunk 5]
```diff
@@ -0,0 +1,36 @@
+<?php
+
+namespace ChoreDo\Migrations;
```
Comment (greydnls) [inline migrations/Version20171023043854.php]: Namespaces should be uniformly camel or lowercase. We've got a bit of each. Maybe do all LC for this PR and then make a separate PR to CC it?

File: src/Entities/Family.php [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+ */
+class Family
+{
+    /**
+     * @var Uuid
+     *
+     * @ORM\Id
+     * @ORM\Column(type="uuid")
+     */
+    private $id;
```
Comment (greydnls) [inline src/Entities/Family.php]: Would highly recommend making a behavior for this. You're going to get tired of typing it.

File: src/Entities/Family.php [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+     */
+    private $weekStartDay;
+
+    /**
+     * @return Uuid
+     */
+    public function getId(): Uuid
+    {
+        if (!isset($this->id)) {
+            $this->id = Uuid::uuid4();
```
Comment (greydnls) [inline src/Entities/Family.php]: o.O

File: src/Entities/Family.php [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+        }
+        return $this->id;
+    }
+
+    /**
+     * @param Uuid $id
+     * @return Family
+     */
+    public function setId(Uuid $id): Family
```
Comment (greydnls) [inline src/Entities/Family.php]: Personally I very much dislike setters and prefer constructor instantiation. It allows for less chance for an invalid/partial state for the objects.

File: src/Entities/Family.php [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+        return $this->paymentStrategy;
+    }
+
+    /**
+     * @param string $paymentStrategy
+     * @return Family
+     */
+    public function setPaymentStrategy(string $paymentStrategy): Family
+    {
+        $this->paymentStrategy = $paymentStrategy;
```
Comment (greydnls) [inline src/Entities/Family.php]: If we're doing a setter on this there needs to be validation. I recommend Beberlei/assert

**Review by greydnls — CHANGES_REQUESTED** [review]

File: cli-config.php (hunk 2 above)
Comment (ieatkillerbees) [reply cli-config.php:60]: Baleeted duplicate code

File: migrations/Version20171023043854.php [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+{
+    /**
+     * @param Schema $schema
+     */
+    public function up(Schema $schema)
+    {
+        // this up() migration is auto-generated, please modify it to your needs
+        $this->abortIf($this->connection->getDatabasePlatform()->getName() !== 'postgresql', 'Migration can only be executed safely on \'postgresql\'.');
+
+        $this->addSql('CREATE TABLE Family (id UUID NOT NULL, name VARCHAR(255) NOT NULL, payment_strategy VARCHAR(255) NOT NULL, completion_threshold INT DEFAULT NULL, week_start_day SMALLINT NOT NULL, PRIMARY KEY(id))');
```
Comment (greydnls) [inline migrations/Version20171023043854.php]: Shouldn't this be pluralized?

File: src/Entities/Behaviors/HasCreatedDate.php [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+     *     name="created",
+     *     nullable=false
+     * )
+     *
+     * @var \DateTime
+     */
+    private $createdDate;
+
+    /** @ORM\PrePersist */
+    public function updateCreatedDate()
```
Comment (greydnls) [inline src/Entities/Behaviors/HasCreatedDate.php:19]: I know you got this from my code, but `updateCreatedDate` sounds weird. What are your thoughts? `setCreatedDate` sounds clearer/less confusing? NBD

File: src/Entities/Behaviors/HasUpdatedDate.php [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+     * )
+     *
+     * @var \DateTime
+     */
+    private $updatedDate;
+
+    /** @ORM\PrePersist */
+    public function updateUpdatedDate()
+    {
+        if ($this->updatedDate === null) {
```
Comment (greydnls) [inline src/Entities/Behaviors/HasUpdatedDate.php]: This isn't accurate. Updated dates should update regardless of being previously set, no?

File: src/Entities/Behaviors/HasUuid.php [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+use Ramsey\Uuid\Uuid;
+
+trait HasUuid
+{
+    /**
+     * @var Uuid
+     *
+     * @ORM\Id
+     * @ORM\Column(type="uuid")
+     * @ORM\GeneratecValue(strategy="UUID")
```
Comment (greydnls) [inline src/Entities/Behaviors/HasUuid.php]: Do you need the Ramsay generator thing or no?

File: src/Entities/Family.php [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+    private $weekStartDay;
+
+    /**
+     * Family constructor.
+     * @param string $name
+     * @param string $paymentStrategy
+     * @param int $completionThreshold
+     * @param int $weekStartDay
+     */
+    public function __construct(string $name, string $paymentStrategy, int $completionThreshold, int $weekStartDay)
```
Comment (greydnls) [inline src/Entities/Family.php]: Completion Threshold is nullable, so it should be nullable here.

Comment (greydnls) [inline src/Entities/Family.php]: weekStartDay is type hinted here as Int but returned as string below.

Comment (greydnls) [inline src/Entities/Family.php]: If they choose "per child" should we require a threshold or possibly default?

File: tests/Entities/Choredo/Tests/FamilyTest.php [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+
+class FamilyTest extends TestCase
+{
+
+    /**
+     * @dataProvider weekDaysProvider
+     */
+    public function testGetStartDayOfWeekReturnsCorrectText($textDay)
+    {
+        $family = new Family();
```
Comment (greydnls) [inline tests/Entities/Choredo/Tests/FamilyTest.php]: You removed the transformation code... how/why is this still passing?

**Review by greydnls — CHANGES_REQUESTED** [review]

File: src/Entities/Behaviors/HasUuid.php [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+     * @ORM\Id
+     * @ORM\Column(type="uuid")
+     * @ORM\GeneratedValue(strategy="UUID")
+     */
+    private $id;
+
+    /**
+     * @return Uuid
+     */
+    public function getId(): Uuid
```
Comment (greydnls) [inline src/Entities/Behaviors/HasUuid.php]: How do you feel about using UuidInterface?

**Review by greydnls — DISMISSED** [review]
